torch/csrc/jit/ir/alias_analysis.cpp

```diff
@@ -7,53 +7,114 @@
-      if (auto elem = getMutableTypePtr(type->cast<FutureType>()->getElementType())) {
-        return FutureType::create(*elem);
+// This class helps convert types to their mutable equivalent by looking up
+// cached conversions.
+class MutableTypePtrHelper {
+ public:
+  explicit MutableTypePtrHelper(
+      std::unordered_map<TypePtr, TypePtr>* mutable_type_cache)
+      : mutable_type_cache_(mutable_type_cache) {}
+
+  c10::optional<TypePtr> getMutableType(const TypePtr& type) {
+    if (mutable_type_cache_) {
+      auto maybe_type = mutable_type_cache_->find(type);
+      if (maybe_type != mutable_type_cache_->end()) {
+        return maybe_type->second;
       }
-      return c10::nullopt;
     }
-    case TypeKind::TupleType: {
-      std::vector<TypePtr> mutable_types;
-      for (const auto& elem : type->expect<TupleType>()->elements()) {
-        if (auto mut_elem = getMutableTypePtr(elem)) {
-          mutable_types.push_back(*mut_elem);
+    auto mutable_type = getMutableTypeImpl(type);
+    if (mutable_type_cache_ && mutable_type) {
+      mutable_type_cache_->emplace(type, *mutable_type);
+    }
+    return mutable_type;
+  }
+
+ private:
+  c10::optional<TypePtr> getMutableTypeImpl(const TypePtr& type) {
+    switch (type->kind()) {
+      case TypeKind::ListType:
+      case TypeKind::DictType:
+      case TypeKind::ClassType:
+      case TypeKind::TensorType:
+        // TODO: lookup cached contained types. this is kind of tricky
+        // because a List[Optional[T]] should still be
+        // List[Optional[Unshaped(T)]], however the getMutableType(Optional[T])
+        // == T
+        return unshapedType(type);
+      case TypeKind::OptionalType:
+        return getMutableType(type->cast<OptionalType>()->getElementType());
+      case TypeKind::FutureType: {
+        if (auto elem =
+                getMutableType(type->cast<FutureType>()->getElementType())) {
+          return FutureType::create(*elem);
         }
-      }
-      if (mutable_types.size() == 0) {
         return c10::nullopt;
-      } else {
-        return TupleType::create(mutable_types);
       }
+      case TypeKind::TupleType: {
+        std::vector<TypePtr> mutable_types;
+        for (const auto& elem : type->expect<TupleType>()->elements()) {
+          if (auto mut_elem = getMutableType(elem)) {
+            mutable_types.push_back(*mut_elem);
+          }
+        }
+        if (mutable_types.size() == 0) {
+          return c10::nullopt;
+        } else {
+          return TupleType::create(mutable_types);
+        }
+      }
+      default:
+        return c10::nullopt;
     }
-    default:
-      return c10::nullopt;
   }
+  std::unordered_map<TypePtr, TypePtr>* mutable_type_cache_;
+};
+
+bool isMutableTypeImpl(
+    const TypePtr& type,
+    std::unordered_map<TypePtr, TypePtr>* mutable_type_cache) {
+  // check common cases to avoid recursively constructing type in
+  // getMutableTypePtrImpl
+  auto kind = type->kind();
```

Comment:
maybe this should be inlined into the mutableTypePtrHelper to avoid an extra indirection?